### ISSUE TITLE
Improve representation of term application ?

### ISSUE THREAD

**Issue by fblanqui** [Deducteam/lambdapi#44]
In many cases, we need to reason about the application head of a term t, that is, the term h such that t = h t1 .. tn and h is not an application. Currently, we need to use the function get_args. We could get it for free by replacing:
  | Appl of term * term
by:
  | Appl of term * term list
where the first component is never an Appl node, and the second component is never empty. This can be enforced by declaring term as a private data type and providing construction functions.
I think that this would clarify some parts of the code and increase the efficiency.
This is done like this in Dedukti.

**Comment by rlepigre** [comment]
I think that we could try, but I'm not so sure that this will buy us much performances. In any case, it would be better to enforce the invariant in the type. For example by defining terms as follows.
```
type term = term_aux * stack
and stack = term list
and term_aux =
  | ... (* usual constructors, without Appl. *)
```
However, I'm not completely sure how this would interact with `Bindlib` (what type for variables and binders?). I'll need to experiment a bit to be sure.

In any case, we need to fix our memory issues before we can do proper benchmarking. So I would suggest that we wait before attempting what you propose.

**Comment by rlepigre** [comment]
I tried several different approaches, and none of them were worth it. Of course, you have a faster implementation of `Basics.get_args`, but at the same time maintaining the representation is costly.

One of the things I tried was to have an application node with the type `term * term array` (which I is what Coq does if I'm not mistaken), and I observed a slowdown. The code was nicer in some parts, but uglier in some others. One problem with working with arrays is that you tend to allocate more, and then the GC is stressed more. So you gain some efficiency, but immediately lose it due to the GC.

Note that the approach used in Dedukti is mostly useless because it uses something like `term * term list`, so the representation contains as many pointer indirections as having binary application nodes. Of course it allows getting the head of a term faster, but it requires many list concatenations to maintain the representation. One could probably fix that using modified lists to have constant time concatenation, but maintaining that would probably have problems as well.

So well, I would suggest to close this issue for now.
